src/python/pants/backend/python/rules/coverage.py

```diff
@@ -0,0 +1,279 @@
+        prefix=address.to_address().path_safe_spec,
+      )
+    )
+    for address, test_result in test_results if test_result._python_sqlite_coverage_file is not None
+  )
+
+  chrooted_sources = await Get[ChrootedPythonSources](HydratedTargets(transitive_targets.closure))
+
+  merged_input_files: Digest = await Get(
+    Digest,
+    DirectoriesToMerge(directories=(
+      *coverage_directory_digests,
+      coveragerc_digest,
+      coverage_setup.requirements_pex.directory_digest,
+      chrooted_sources.digest,
+    )),
+  )
+
+  prefixes = [f'{address.to_address().path_safe_spec}/.coverage' for address, _ in test_results]
+  coverage_args = ['combine', *prefixes]
+  request = coverage_setup.requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path=f'./{coverage_setup.filename}',
+    pex_args=coverage_args,
+    input_files=merged_input_files,
+    output_files=('.coverage',),
+    description=f'Merge coverage reports.',
+  )
+
+  result = await Get[ExecuteProcessResult](
+    ExecuteProcessRequest,
+    request
+  )
+  return MergedCoverageData(coverage_data=result.output_directory_digest)
+
+
+COVERAGE_PLUGIN_MODULE_NAME = '__coverage_coverage_plugin__'
+
+
+def get_coverage_plugin_input():
+  return InputFilesContent(
+    FilesContent(
+      (
+        FileContent(
+          path=f'{COVERAGE_PLUGIN_MODULE_NAME}.py',
+          content=pkg_resources.resource_string(__name__, 'coverage/plugin.py'),
+          is_executable=False,
+        ),
+      )
+    )
+  )
+
+
+def ensure_section(config_parser: configparser.ConfigParser, section: str) -> None:
+  """Ensure a section exists in a ConfigParser."""
+  if not config_parser.has_section(section):
+    config_parser.add_section(section)
+
+
+def construct_coverage_config(source_roots, python_files) -> str:
+  # A map from source root stripped source to its source root. eg:
+  #  {'pants/testutil/subsystem/util.py': 'src/python'}
+  # This is so coverage reports referencing /chroot/path/pants/testutil/subsystem/util.py can be mapped
+  # back to the actual sources they reference when merging coverage reports.
+  source_to_target_base: Dict[str, str] = {}
+  for file_name in python_files:
+    source_root = source_roots.find_by_path(file_name)
+    source_root_stripped_path = file_name[len(source_root.path) + 1:]
+    source_to_target_base[source_root_stripped_path] = source_root.path
+
+  config_parser = configparser.ConfigParser()
+  config_parser.read_file(StringIO(DEFAULT_COVERAGE_CONFIG))
+  ensure_section(config_parser, 'run')
+  config_parser.set('run', 'plugins', COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.add_section(COVERAGE_PLUGIN_MODULE_NAME)
+  config_parser.set(COVERAGE_PLUGIN_MODULE_NAME, 'source_to_target_base', json.dumps(source_to_target_base))
+  config = StringIO()
+  config_parser.write(config)
+  return config.getvalue()
+
+
+def get_file_names(all_target_adaptors):
```

Comment:
Would you mind adding a TODO here referencing #4535? This is a basic facility that it should provide.